File: get_message_code_server.py
# -*- coding: utf-8 -*-
import sqlite3
import os
import re
from fastapi import FastAPI
import uvicorn

app = FastAPI()
DB_PATH = os.path.expanduser("~/Library/Messages/chat.db")


def get_latest_verification_code():
    # 使用上下文管理器, 确保连接及时关闭
    with sqlite3.connect(DB_PATH) as conn:
        cursor = conn.cursor()
        query = """
        SELECT text
        FROM message
        WHERE datetime(date/1000000000 + 978307200, "unixepoch", "localtime") > datetime("now", "localtime", "-60 second")
        ORDER BY date DESC
        LIMIT 1;
        """
        cursor.execute(query)
        result = cursor.fetchone()
        if result:
            text = result[0]
            # 检查消息中是否包含“验证码”或其他常见关键词
            if re.search(r"验证码|校验码|动态码|短信码", text):
                # 提取 4 到 6 位数字的验证码
                match = re.search(r"\d{4,6}", text)
                if match:
                    code = match.group()
                    return {"sms_code": code, "message": "Code found"}
                else:
                    return {"sms_code": -1, "message": "No code found"}
            else:
                return {"sms_code": -1, "message": "No code found"}
        else:
            return {"sms_code": -1, "message": "No message found"}


@app.get("/get_code")
async def get_code():
    result = get_latest_verification_code()
    return result


if __name__ == "__main__":
    uvicorn.run(app, host="127.0.0.1", port=65530)

File: get_message_code_server.go
package main

import (
	"database/sql"
	"fmt"
	"github.com/gin-gonic/gin"
	"os"
	"path/filepath"
	"regexp"

	_ "github.com/mattn/go-sqlite3"
	"net/http"
)

func getLatestVerificationCode() gin.H {
	// 获取数据库路径
	homeDir, _ := os.UserHomeDir()
	dbPath := filepath.Join(homeDir, "Library", "Messages","chat.db")

	// 检查数据库文件是否存在
	if _, err := os.Stat(dbPath); os.IsNotExist(err) {
		return gin.H{"sms_code": -1, "message": "Database file not found"}
	}

	// 连接数据库
	db, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return gin.H{"sms_code": -1, "message": "Database connection error"}
	}
	defer db.Close()

	// 构建查询语句
	query := `
    SELECT text
    FROM message
    WHERE datetime(date/1000000000 + 978307200, 'unixepoch', 'localtime') > datetime('now', 'localtime', '-60 seconds')
    ORDER BY date DESC
    LIMIT 1;
    `

	// 执行查询
	row := db.QueryRow(query)
	var text string
	err = row.Scan(&text)
	if err != nil {
		if err == sql.ErrNoRows {
			return gin.H{"sms_code": -1, "message": "No message found"}
		}
		return gin.H{"sms_code": -1, "message": "Database query error"}
	}

	// 检查关键词
	keywordRegex := regexp.MustCompile(`验证码|校验码|动态码|短信码`)
	if keywordRegex.MatchString(text) {
		// 提取验证码
		codeRegex := regexp.MustCompile(`\d{4,6}`)
		code := codeRegex.FindString(text)
		if code != "" {
			return gin.H{"sms_code": code, "message": "Code found"}
		} else {
			return gin.H{"sms_code": -1, "message": "No code found"}
		}
	} else {
		return gin.H{"sms_code": -1, "message": "No code found"}
	}
}

func main() {
	gin.SetMode(gin.ReleaseMode)
	router := gin.Default() // 创建一个默认的路由引擎

	router.GET("/get_code", func(c *gin.Context) {
		result := getLatestVerificationCode()
		c.JSON(http.StatusOK, result)
	})

	// 监听端口65530
	err := router.Run("127.0.0.1:65530")
	if err != nil {
		fmt.Println("Failed to start server:", err)
	}
}
